blockchain/subsidy.go

```diff
@@ -26,36 +64,63 @@ import (
 // 2     subsidy /= DivSubsidy
 //
 // Safe for concurrent access.
-func calcBlockSubsidy(height int64, params *chaincfg.Params) int64 {
+func (s *SubsidyCache) CalcBlockSubsidy(height int64) int64 {
 	// Block height 1 subsidy is 'special' and used to
 	// distribute initial tokens, if any.
 	if height == 1 {
-		return params.BlockOneSubsidy()
+		return s.params.BlockOneSubsidy()
 	}
 
-	iterations := height / params.ReductionInterval
-	subsidy := params.BaseSubsidy
-
-	// You could stick all these values in a LUT for faster access if you
-	// wanted to, but this calculation is already really fast until you
-	// get very very far into the blockchain. The other method you could
-	// use is storing the total subsidy in a block node and do the
-	// multiplication and division when needed when adding a block.
-	if iterations > 0 {
-		for i := int64(0); i < iterations; i++ {
-			subsidy *= params.MulSubsidy
-			subsidy /= params.DivSubsidy
-		}
+	iteration := uint64(height / s.params.ReductionInterval)
+
+	if iteration == 0 {
+		return s.params.BaseSubsidy
+	}
+
+	// First, check the cache.
+	s.subsidyCacheLock.RLock()
+	if cachedValue, existsInCache := s.subsidyCache[iteration]; existsInCache {
+		subsidy := cachedValue
```

Comment:
`cachedValue` is already a copy, so there is no need to copy it again.  You can just delete this and do:

```Go
	s.subsidyCacheLock.RUnlock()
	return cachedValue
```